### PR TITLE
Add Cross Navigation Enrichment

### PR DESCRIPTION
**Pull request by adatzer**
This pull request adds an enrichment that can parse the extended cross navigation format in `_sp` querystring parameter and attach the `cross_navigation` context to an event.

The extended cross navigation format can be described by `_sp={domainUserId}.{timestamp}.{sessionId}.{subjectUserId}.{sourceId}.{platform}.{reason}`

The changes in the trackers to allow this extended format can be seen in:

- [Android tracker](https://github.com/snowplow/snowplow-android-tracker/pull/640) 
- [JavaScript tracker](https://github.com/snowplow/snowplow-javascript-tracker/pull/1233) 

The corresponding iglu-central pull requests are:

- [to add the `cross_navigation` context schema](https://github.com/snowplow/iglu-central/pull/1327)
- [to add the `cross_navigation_config` enrichment schema](https://github.com/snowplow/iglu-central/pull/1347)

cc @greg-el @igneel64 @matus-tomlein

### CODE REVIEW

**Review by matus-tomlein — COMMENTED** [review]
Looks good! Will try to test this with the tracker implementations.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/web/PageEnrichments.scala [hunk 1]
```diff
@@ -52,19 +53,12 @@ object PageEnrichments {
    * @param qsMap The querystring parameters
    * @return Validation boxing a pair of optional strings corresponding to the two fields
    */
-  def parseCrossDomain(qsMap: QueryStringParameters): Either[FailureDetails.EnrichmentFailure, (Option[String], Option[String])] =
+  def parseCrossDomain(qsMap: QueryStringParameters): Either[FailureDetails.EnrichmentFailure, Map[String, Option[String]]] =
```
[the remark this answers is not in the document]
Comment (adatzer) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/web/PageEnrichments.scala:56]: I had some back and forths on this. Happy to move this over completely, it just felt that this function parses querystring params and the fact that it deals only with `_sp` is "coincidence", e.g. maybe in the future it deals with more. What do you think?

**Comment by matus-tomlein** [comment]
Just tested with the mobile trackers implementation, it works nicely, great work! Will do a test with the JS tracker tomorrow.

File: config/enrichments/cross_navigation_config.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+    "schema": "iglu:com.snowplowanalytics.snowplow.enrichments/cross_navigation_config/jsonschema/1-0-0",
+
+    "data": {
+        "enabled": false,
```
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply config/enrichments/cross_navigation_config.json:5]: Nevermind the part about the flow, we can get better feedback from DP on this. But I do think that this configuration suggests the default – at least when I ran enrich locally, I had to change it to true in order for the new behaviour to work.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/web/PageEnrichments.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/web/PageEnrichments.scala:56]: Ah I see your point, but in case we have more query parameters to parse, wouldn't we have multiple functions for parsing each one separately? Not sure I see the value of having it in one, but this is just a nitpick, this approach sounds fine as well.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/CrossNavigationEnrichment.scala [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+   */
+  def makeCrossDomainMap(sp: String): Either[FailureDetails.EnrichmentFailure, Map[String, Option[String]]] =
+    sp.split("\\.", -1)
+      .padTo(
+        CrossNavProps.size,
+        ""
+      )
+      .toList match {
+      case List(
+            duidElt,
```
Comment (spenes) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/CrossNavigationEnrichment.scala:148]: It seems like there is too much repetition here. I think we can reduce the repetition by putting transformation functions and props name together. I've implemented what had in mind [here](https://github.com/snowplow/enrich/tree/feat/cross_navigation_cleanup). Let me know if you see any problem with it.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala [hunk 4]
```diff
@@ -607,18 +607,35 @@ object EnrichmentManager {
...
+              .parseCrossDomain(qsMap)
+              .bimap(
+                err =>
+                  crossNavEnrichment match {
+                    case Some(cn) => NonEmptyList.one(cn.addEnrichmentInfo(err))
+                    case None => NonEmptyList.one(err)
+                  },
+                crossNavMap => {
+                  val crossNavKeys = crossNavMap.keySet
+                  val duidKey = CNE.CrossNavProps(0)
```
Comment (spenes) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:627]: This seems a bit unsafe and hard to follow as it is. I think it would better if we refactor out `duidKey` to some variable in `CrossNavigationEnrichment` and use that instead.

Or, we can solve this problem with wrapping `crossNavMap` to some class such as `CrossDomainMap`. With this way, we wouldn't have to bother witn map itself or its key names. It can be abstracted out to a method in `CrossDomainMap` class.

I've implemented what I had in mind in [here](https://github.com/snowplow/enrich/tree/feat/cross_navigation_cleanup).

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/CrossNavigationEnrichment.scala [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+   */
+  private def reformatTstamp(tstamp: Option[String]): Option[String] = {
+    val pFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
+    val formatter = DateTimeFormatter.ISO_DATE_TIME
+
+    tstamp match {
+      case Some(t) =>
+        Some(formatter.format(pFormatter.parse(t)).replaceAll(" ", "T") + "Z")
+      case None => None
+    }
```
Comment (spenes) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/CrossNavigationEnrichment.scala:119]:
```suggestion
    tstamp.map(t => formatter.format(pFormatter.parse(t)).replaceAll(" ", "T") + "Z")
```

**Review by spenes — CHANGES_REQUESTED** [review]
Thanks @adatzer! The changes looks good in general. I raised a question regarding using feature flag to enable/disable feature instead of implementing it as a new enrichment. Just ignore it if it is already discussed.

Also, I suggested some structural changes to make code more concise and a bit less error prone. I implemented the changes I had in mind in [here](https://github.com/snowplow/enrich/tree/feat/cross_navigation_cleanup) to reduce the back and forth. Feel free to challenge any changes in there!

File: config/enrichments/cross_navigation_config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (spenes) [reply config/enrichments/cross_navigation_config.json:5]: On second thought, I am in favor of making it enrichment because it isn’t just some internal change in the Enrich. It adds new context to events therefore it is actually ‘enriching’ the event therefore it would be better to make it enrichment. You can ignore above comment.

**Comment by adatzer** [comment]
Hi Enes! Thank you very much for reviewing and for creating the cleanup branch. This was really helpful!
I have cherry-picked your commit as is, and also applied scalafmt.

The additional `Make edits` commit includes some edits related to:

1. The `cross-navigation` schema ([pending final review](https://github.com/snowplow/iglu-central/pull/1327)). Since the `timestamp` and the `domain_user_id` props are required, we decided that if they are missing, the enrichment should not add the corresponding entity at all, i.e. this is considered invalid extended format.
2. As discussed with Analytics Engineers it is a bit preferrable for optional properties to be null (which is valid according to schema), rather than missing. This is why the `filterNot` was removed.
3. Corresponding test changes related to the above.

Additionally, this PR has 2 iglu-central pull requests as dependencies:

- https://github.com/snowplow/iglu-central/pull/1327 (adds the context schema)
- https://github.com/snowplow/iglu-central/pull/1347 (adds the config schema)

Please let us know what you think of these last changes and whether there is anything else to be done from our side. Thanks once again!

cc @igneel64 @greg-el @matus-tomlein

**Review by spenes — APPROVED** [review]
Thanks Ada, the new changes look good to me!
